libs/community/langchain_community/llms/vertexai.py

```diff
@@ -309,11 +338,89 @@ def _generate(
+                    [self._response_to_generation(r) for r in res.candidates]
                 )
-                generations.append([_response_to_generation(r) for r in res.candidates])
         return LLMResult(generations=generations)
 
+    def invoke(
+        self,
+        input: LanguageModelInput,
+        config: Optional[RunnableConfig] = None,
+        *,
+        stop: Optional[List[str]] = None,
+        **kwargs: Any,
+    ) -> str:
+        vertex_input = self._convert_to_prompt(model_input=input)
+        if isinstance(vertex_input, str):
+            return super().invoke(input, config, stop=stop, **kwargs)
+        if not isinstance(vertex_input, list):
+            raise ValueError(
+                "Input for a multimodal model should be a str or a list of elements!"
+            )
+        params = self._prepare_params(stop=stop, stream=False, **kwargs)
+        res = completion_with_retry(
+            self,
+            vertex_input,
+            is_gemini=True,
+            run_manager=kwargs.get("run_manager"),
+            **params,
+        )
+
+        return res.text
+
+    def _convert_to_prompt(
+        self, model_input: LanguageModelInput
+    ) -> Union[str, StringPromptValue, List[Union[str, "Image"]]]:
+        if isinstance(model_input, str):
+            return model_input
+        elif isinstance(model_input, StringPromptValue):
+            return model_input.to_string()
+        elif isinstance(model_input, list):
+            if "vision" not in self.model_name:
+                raise ValueError(
+                    f"Multi-modal input is not supported by {self.model_name} model!"
+                )
+            first_message = model_input[0]
+            if len(model_input) > 1 and not isinstance(first_message, BaseMessage):
+                raise ValueError(
+                    "Multi-modal model expects only a single message as a input!"
+                )
+        else:
+            raise ValueError(
+                f"Invalid input type {type(model_input)}. "
+                "Must be a PromptValue, str, or list of BaseMessages."
+            )
+
+        from vertexai.preview.generative_models import Image
+
+        vertex_messages: List[Union[str, Image]] = []
+        for content in first_message.content:
+            if not isinstance(content, Dict):
+                raise ValueError(
+                    f"Message's content is expected to be a dict, got {type(content)}!"
+                )
+            if content["type"] == "text":
+                vertex_messages.append(content["text"])
+            elif content["type"] == "image_url":
+                path = content["image_url"]["url"]
+                if path.startswith("gs://"):
+                    vertex_messages.append(
+                        _load_image_from_gcs(path=path, project=self.project)
+                    )
+                else:
+                    vertex_messages.append(Image.load_from_file(path))
+            else:
```

Comment:
It would be good to find a way to allow passing in an image as `bytes` instead of only allowing loading from a file or from GCS. Unfortunately I can't think of a good way to distinguish that case while staying strictly compatible with the OpenAI message format. If you're willing to extend the format, perhaps you could add a case like:

```
elif content["type"] == "image":
  vertex_messages.append(Image.from_bytes(content["image"])
```

(or I can send this as a followup PR)

Reply:
done